mongodb/transaction_test.go

```diff
@@ -0,0 +1,255 @@
+	}
+	conn, err := mongoDriver.Open(driverConfig)
+	if err != nil {
+		// log error, cannot use mongo db
+	}
+
+	r, e := conn.RunTransaction(ctx, true, exampleTransactionFunc(conn, false))
+
+	switch {
+	// handle this special case, where we have aborted the transaction because the object was not in a valid state
+	case errors.Is(e, badObjectState):
+
+	// otherwise, a runtime error, i.e. couldn't complete the transaction for some other reason (even with retries)
+	case !errors.Is(e, nil):
+
+	// transaction completed successfully, and r contains a valid object containing the 2 returned simpleObjects
+	default:
+		if _, ok := r.(struct{ o1, o2 simpleObject }); !ok {
+			// Armageddon!
+		}
+	}
+}
+
+var badObjectState = errors.New("object state incorrect")
+
+// exampleTransactionFunc returns a sample TransactionFunc that executes a series of mongo based operations
+// within a transaction defined by the transactionCtx parameter of the function.
+// The transactionCtx is derived from the MongoConnection conn
+// The interleave boolean value allows injection a mongo call outside of the main transaction in transactionCtx, to
+// simulate a parallel transaction that interferes with the main transaction. Without retrying the main transaction,
+// this will lead to a Transaction Error on commit
+func exampleTransactionFunc(conn *mongoDriver.MongoConnection, interleave bool) mongoDriver.TransactionFunc {
+	return func(transactionCtx context.Context) (interface{}, error) {
+		var obj1, obj2 simpleObject
+		err := conn.Collection(collection1).FindOne(transactionCtx, bson.M{"_id": 1}, &obj1)
+		if err != nil {
+			return nil, fmt.Errorf("could not find object in collection (%s): %w", collection1, err)
+		}
+
+		switch obj1.State {
+		case "first":
+			obj1.State = "second"
+		case "second":
+			obj1.State = "third"
+
+			obj2 = simpleObject{ID: 1, State: "final"}
+			_, err = conn.Collection(collection2).Upsert(transactionCtx, bson.M{"_id": obj2.ID}, bson.M{"$set": obj2})
+			if err != nil {
+				return nil, fmt.Errorf("could not upsert object in collection (%s): %w", collection2, err)
+			}
+		default:
+			return nil, badObjectState
+		}
+
+		if interleave {
+			_, err = conn.Collection(collection1).Update(context.Background(), bson.M{"_id": 1}, bson.M{"$set": bson.M{"state": "second"}})
+			if err != nil {
+				return nil, fmt.Errorf("interleave write failed in collection (%s): %w", collection1, err)
+			}
+		}
+
+		_, err = conn.Collection(collection1).Update(transactionCtx, bson.M{"_id": 1}, bson.M{"$set": obj1})
+		if err != nil {
+			return nil, fmt.Errorf("could not write object in collection (%s): %w", collection1, err)
+		}
+
+		return struct{ o1, o2 simpleObject }{o1: obj1, o2: obj2}, nil
+	}
+}
+
+func TestTransaction(t *testing.T) {
+	ctx := context.Background()
+	conn, cleanup := setupMongoConnection(t)
+	defer cleanup(ctx)
+
+	Convey("Given a mongo server cluster", t, func() {
```

Comment:
Can we move the setup lines to this convey?

Reply:
yes but the the setupMongoConnection() will then be called for each of the Conveys on 109, 154, etc. Unless you are suggesting doing that and setting up according to what is required for each  those individual Conveys?

Reply:
Done